database/misc/elasticsearch_shield_role.py

```diff
@@ -0,0 +1,236 @@
+        return resp, info
+
+    def get_role(self, name):
+        url = urlparse.urljoin(self.elasticsearch_api, '/_shield/role/%s' % name)
+        resp, info = self._do_request(url, 'GET')
+        if int(info['status']) == 200:
+            try:
+                content = json.loads(resp.read())
+            except AttributeError:
+                # there was no content, but the error read()
+                # may have been stored in the info as 'body'
+                content = json.loads(info.pop('body', ''))
+
+            return True, content[name]
+        
+        # role doesn't exist
+        elif int(info['status']) == 404:
+            return False, None
+        
+        self.module.fail_json(msg=info)
+
+
+    def role_del(self, name):
+        url = urlparse.urljoin(self.elasticsearch_api, '/_shield/role/%s' % name)
+        resp, info = self._do_request(url, 'DELETE')
+        if int(info['status']) == 200:
+            return True
+        
+        self.module.fail_json(msg=info)
+
+    def role_add(self, name, cluster, indices, run_as):
+        url = urlparse.urljoin(self.elasticsearch_api, '/_shield/role/%s' % name)
+        body = {}
+        
+        body['cluster'] = cluster
+        body['indices'] = indices
+        if run_as:
+            body['run_as'] = run_as
+        
+        resp, info = self._do_request(url, 'POST', body)
+
+        if int(info['status']) == 200:
+            return True
+        
+        self.module.fail_json(msg=info)
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            name=dict(required=True),
+            cluster=dict(required=False, type='list'),
+            indices=dict(required=False, type='list'),
+            run_as=dict(required=False, default=None, type='list'),
+            url_username = dict(required=False, default=None, aliases=['admin_user']),
+            url_password = dict(required=False, default=None, no_log=True, aliases=['admin_password']),
+            state=dict(default='present', choices=['present', 'absent']),
+            elasticsearch_api=dict(default='http://localhost:9200'),
+            update_role=dict(default='always', choices=['always', 'on_create']),
+        )
+    )
+
+    name                = module.params['name']
+    cluster             = module.params['cluster']
+    indices             = module.params['indices']
+    run_as              = module.params['run_as']
+    state               = module.params['state']
+    elasticsearch_api   = module.params['elasticsearch_api']
+    update_role         = module.params['update_role']
+
+    shield_role = ShieldNativeRole(module, elasticsearch_api)
+
+    present, role = shield_role.get_role(name)
+
+    if state == 'present':
+        if not cluster or not indices:
+            module.fail_json(msg="cluster and indices are required to add role")
+
+        if present:
+            if update_role == 'always':
```

Comment:
It seems you always update the role no matter if the existing data is different. right? Is there a way to just compare the existing data and only update the role if required?